### PR TITLE
Surgical multi-relationship removal - Addresses FCREPO-1362

### PR DESCRIPTION
**Pull request by whikloj**
See: [FCREPO-1362](https://jira.duraspace.org/browse/FCREPO-1362)

### CODE REVIEW

File: fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/FedoraResourceImpl.java [hunk 1]
```diff
@@ -257,7 +260,22 @@ public void delete() {
             final Iterator<Property> inboundProperties = Iterators.concat(references, weakReferences);
 
             while (inboundProperties.hasNext()) {
-                inboundProperties.next().remove();
+                final Property prop = inboundProperties.next();
+                final List<Value> newVals = new ArrayList<Value>();
```
Comment (ajs6f) [inline fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/FedoraResourceImpl.java]: Use the diamond operator: `new ArrayList<>()`.

File: fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/FedoraResourceImpl.java [hunk 2]
```diff
@@ -257,7 +260,22 @@ public void delete() {
...
+                final Iterator<Value> propIt = JcrPropertyFunctions.property2values.apply(prop);
```
Comment (ajs6f) [inline fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/FedoraResourceImpl.java]: Use a static import.

File: fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/FedoraResourceImpl.java [hunk 3]
```diff
@@ -257,7 +260,22 @@ public void delete() {
...
+                while (propIt.hasNext()) {
+                    final Value v = propIt.next();
+                    final Node n = this.getSession().getNodeByIdentifier(v.getString());
```
Comment (ajs6f) [inline fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/FedoraResourceImpl.java]: You can drop the `this` for just `getSession()`.

File: fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/FedoraResourceImpl.java [hunk 4]
```diff
@@ -257,7 +260,22 @@ public void delete() {
             final Iterator<Property> inboundProperties = Iterators.concat(references, weakReferences);
 
             while (inboundProperties.hasNext()) {
-                inboundProperties.next().remove();
+                final Property prop = inboundProperties.next();
+                final List<Value> newVals = new ArrayList<Value>();
+                final Iterator<Value> propIt = JcrPropertyFunctions.property2values.apply(prop);
+                while (propIt.hasNext()) {
```
Comment (ajs6f) [inline fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/FedoraResourceImpl.java]: Given how few references there are likely to be, you might consider rolling up the iterator into a collection and then `remove`ing the one you don't want.

Comment (whikloj) [inline fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/FedoraResourceImpl.java]: Is that of any benefit? I don't think I can use remove(Object) so I am still iterating over the list. But I might be missing something.

Comment (ajs6f) [inline fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/FedoraResourceImpl.java]: Oh, shoot, you're right. You can't. But you can still do better than this via [Guava's `removeIf`](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/Iterators.html#removeIf%28java.util.Iterator,%20com.google.common.base.Predicate%29).

Comment (whikloj) [inline fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/FedoraResourceImpl.java]: Ok, that is interesting, but I'm not sure that using a Predicate doesn't make the code more confusing.
Also the **property2values** function returns an UnmodifiableIterator, so the **removeIf** fails to modify it.
I put this attempt in a [separate branch](https://github.com/whikloj/fcrepo4/blob/FCREPO-1362-2/fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/FedoraResourceImpl.java#L262-L283) so you can see what I've done, incase I missed something.

Comment (ajs6f) [inline fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/FedoraResourceImpl.java]: Oh, I didn't know that `property2values` did that-- that's lame. It's not worth worrying about this any more anyway- this all goes away with Java 8's Streams, so this is limited-lifetime code. I'm sorry I took up this much time with it.

**Comment by awoods** [comment]
Resolved with: https://github.com/fcrepo4/fcrepo4/commit/4a77e57d18687b6e4f653e0d88765d007e284584

**Pull request closed without merging**